pkg/kubelet/images/pullmanager/fs_pullrecords.go

```diff
@@ -0,0 +1,301 @@
+func NewFSPullRecordsAccessor(kubeletDir string) (*fsPullRecordsAccessor, error) {
+	kubeletConfigEncoder, kubeletConfigDecoder, err := createKubeletConfigSchemeEncoderDecoder()
+	if err != nil {
+		return nil, err
+	}
+
+	accessor := &fsPullRecordsAccessor{
+		pullingDir: filepath.Join(kubeletDir, "image_manager", "pulling"),
+		pulledDir:  filepath.Join(kubeletDir, "image_manager", "pulled"),
+
+		encoder: kubeletConfigEncoder,
+		decoder: kubeletConfigDecoder,
+	}
+
+	if err := os.MkdirAll(accessor.pullingDir, 0700); err != nil {
+		return nil, err
+	}
+
+	if err := os.MkdirAll(accessor.pulledDir, 0700); err != nil {
+		return nil, err
+	}
+
+	return accessor, nil
+}
+
+func (f *fsPullRecordsAccessor) WriteImagePullIntent(image string) error {
+	intent := kubeletconfiginternal.ImagePullIntent{
+		Image: image,
+	}
+
+	intentBytes := bytes.NewBuffer([]byte{})
+	if err := f.encoder.Encode(&intent, intentBytes); err != nil {
+		return err
+	}
+
+	return writeFile(f.pullingDir, cacheFilename(image), intentBytes.Bytes())
+}
+
+func (f *fsPullRecordsAccessor) ListImagePullIntents() ([]*kubeletconfiginternal.ImagePullIntent, error) {
+	var intents []*kubeletconfiginternal.ImagePullIntent
+	// walk the pulling directory for any pull intent records
+	err := processDirFiles(f.pullingDir,
+		func(filePath string, fileContent []byte) error {
+			intent, err := decodeIntent(f.decoder, fileContent)
+			if err != nil {
+				return fmt.Errorf("failed to deserialize content of file %q into ImagePullIntent: %w", filePath, err)
+			}
+			intents = append(intents, intent)
+
+			return nil
+		})
+	return intents, err
+}
+
+func (f *fsPullRecordsAccessor) ImagePullIntentExists(image string) (bool, error) {
+	intentRecordPath := filepath.Join(f.pullingDir, cacheFilename(image))
+	intentBytes, err := os.ReadFile(intentRecordPath)
+	if os.IsNotExist(err) {
+		return false, nil
+	} else if err != nil {
+		return false, err
+	}
+
+	intent, err := decodeIntent(f.decoder, intentBytes)
+	if err != nil {
+		return false, err
+	}
+
+	return intent.Image == image, nil
+}
+
+func (f *fsPullRecordsAccessor) DeleteImagePullIntent(image string) error {
+	err := os.Remove(filepath.Join(f.pullingDir, cacheFilename(image)))
+	if os.IsNotExist(err) {
+		return nil
+	}
+	return err
+}
+
+func (f *fsPullRecordsAccessor) GetImagePulledRecord(imageRef string) (*kubeletconfiginternal.ImagePulledRecord, bool, error) {
+	recordBytes, err := os.ReadFile(filepath.Join(f.pulledDir, cacheFilename(imageRef)))
+	if os.IsNotExist(err) {
+		return nil, false, nil
+	} else if err != nil {
+		return nil, false, err
+	}
+
+	pulledRecord, err := decodePulledRecord(f.decoder, recordBytes)
+	if err != nil {
+		return nil, true, err
+	}
+	if pulledRecord.ImageRef != imageRef {
+		return nil, false, nil
+	}
+	return pulledRecord, true, err
+}
+
+func (f *fsPullRecordsAccessor) ListImagePulledRecords() ([]*kubeletconfiginternal.ImagePulledRecord, error) {
```

Comment:
document that this can return partial success (non-empty records AND an error indicating issues were encountered during list)